File: backend/app/api/routes/mapping_profiles.py
import uuid

from fastapi import APIRouter

from app.schemas.mapping import (
    MappingProfileCreate,
    MappingProfileResponse,
    MappingProfileVersionResponse,
)

router = APIRouter(prefix="/api/mapping-profiles", tags=["mapping-profiles"])

_mapping_profiles: dict[str, MappingProfileResponse] = {}


@router.post("", response_model=MappingProfileResponse)
def create_mapping_profile(payload: MappingProfileCreate) -> MappingProfileResponse:
    profile_id = str(uuid.uuid4())
    response = MappingProfileResponse(
        id=profile_id,
        company_id=payload.company_id,
        name=payload.name,
        bank_template_id=payload.bank_template_id,
        company_journal_template_id=payload.company_journal_template_id,
        status="active",
        latest_version=MappingProfileVersionResponse(version_no=1, **payload.version.model_dump()),
    )
    _mapping_profiles[profile_id] = response
    return response


@router.get("", response_model=list[MappingProfileResponse])
def list_mapping_profiles(company_id: str | None = None) -> list[MappingProfileResponse]:
    profiles = list(_mapping_profiles.values())
    if company_id is not None:
        profiles = [p for p in profiles if p.company_id == company_id]
    return profiles

File: frontend/src/api/mappingProfiles.ts
import { apiClient } from './client';
import type { MappingProfile, MappingProfileVersion } from '../types/mapping';

export async function listMappingProfiles(companyId?: string): Promise<MappingProfile[]> {
  const params = companyId ? { company_id: companyId } : undefined;
  const response = await apiClient.get<MappingProfile[]>('/api/mapping-profiles', { params });
  return response.data;
}

export async function getMappingProfile(id: string): Promise<MappingProfile> {
  const response = await apiClient.get<MappingProfile>(`/api/mapping-profiles/${id}`);
  return response.data;
}

export async function createMappingProfile(payload: {
  company_id: string;
  name: string;
  bank_template_id?: string | null;
  company_journal_template_id?: string | null;
  version: Partial<MappingProfileVersion>;
}): Promise<MappingProfile> {
  const response = await apiClient.post<MappingProfile>('/api/mapping-profiles', payload);
  return response.data;
}

export async function createMappingProfileVersion(
  id: string,
  version: Partial<MappingProfileVersion>
): Promise<MappingProfile> {
  const response = await apiClient.post<MappingProfile>(
    `/api/mapping-profiles/${id}/versions`,
    version
  );
  return response.data;
}

export async function listMappingProfileVersions(
  id: string
): Promise<MappingProfileVersion[]> {
  const response = await apiClient.get<MappingProfileVersion[]>(
    `/api/mapping-profiles/${id}/versions`
  );
  return response.data;
}

export async function setMappingProfileStatus(
  id: string,
  status: 'active' | 'inactive'
): Promise<MappingProfile> {
  const response = await apiClient.patch<MappingProfile>(
    `/api/mapping-profiles/${id}/status`,
    null,
    { params: { status } }
  );
  return response.data;
}
